File: src/electricity_shop_mobile/lib/shared/models/paged_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'paged_response.g.dart';

/// Generic class for paginated API responses
@JsonSerializable(genericArgumentFactories: true)
class PagedResponse<T> {
  /// Current page number
  final int pageNumber;
  
  /// Number of items per page
  final int pageSize;
  
  /// Total number of pages
  final int totalPages;
  
  /// Total number of items across all pages
  final int totalCount;
  
  /// Whether there's a previous page available
  final bool hasPrevious;
  
  /// Whether there's a next page available
  final bool hasNext;
  
  /// Collection of items for the current page
  final List<T> items;

  /// Creates a new [PagedResponse] instance
  PagedResponse({
    required this.pageNumber,
    required this.pageSize,
    required this.totalPages,
    required this.totalCount,
    required this.hasPrevious,
    required this.hasNext,
    required this.items,
  });

  /// Creates a [PagedResponse] from JSON map using a provided JSON converter for T
  factory PagedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? json) fromJsonT,
  ) => _$PagedResponseFromJson(json, fromJsonT);

  /// Converts [PagedResponse] to JSON map using a provided JSON converter for T
  Map<String, dynamic> toJson(Object? Function(T value) toJsonT) => 
      _$PagedResponseToJson(this, toJsonT);
}

File: src/ElectricityShop.API/Models/PagedResponse.cs
using System.Collections.Generic;

namespace ElectricityShop.API.Models
{
    /// <summary>
    /// Generic wrapper for paginated API responses
    /// </summary>
    /// <typeparam name="T">Type of items in the collection</typeparam>
    public class PagedResponse<T>
    {
        /// <summary>
        /// Current page number
        /// </summary>
        public int PageNumber { get; set; }

        /// <summary>
        /// Number of items per page
        /// </summary>
        public int PageSize { get; set; }

        /// <summary>
        /// Total number of pages
        /// </summary>
        public int TotalPages { get; set; }

        /// <summary>
        /// Total number of items across all pages
        /// </summary>
        public int TotalCount { get; set; }

        /// <summary>
        /// Whether there's a previous page available
        /// </summary>
        public bool HasPrevious => PageNumber > 1;

        /// <summary>
        /// Whether there's a next page available
        /// </summary>
        public bool HasNext => PageNumber < TotalPages;

        /// <summary>
        /// The collection of items for the current page
        /// </summary>
        public IList<T> Items { get; set; }

        public PagedResponse()
        {
            Items = new List<T>();
        }

        public PagedResponse(IList<T> items, int count, int pageNumber, int pageSize)
        {
            PageNumber = pageNumber;
            PageSize = pageSize;
            TotalCount = count;
            TotalPages = (int)Math.Ceiling(count / (double)pageSize);
            Items = items;
        }
    }
}